src/mongocxx/instance.cpp

```diff
@@ -14,31 +14,73 @@
 
 #include <mongocxx/instance.hpp>
 
+#include <mutex>
+#include <utility>
+
 #include <bsoncxx/stdx/make_unique.hpp>
 
-#include <mongoc.h>
+#include <mongocxx/logger.hpp>
+#include <mongocxx/private/libmongoc.hpp>
+
+namespace mongocxx {
+MONGOCXX_INLINE_NAMESPACE_BEGIN
 
 namespace {
-void log_handler(mongoc_log_level_t, const char *, const char *, void *) {
+
+log_level convert_log_level(::mongoc_log_level_t mongoc_log_level) {
+    switch (mongoc_log_level) {
+        case MONGOC_LOG_LEVEL_ERROR:
+            return log_level::k_error;
+        case MONGOC_LOG_LEVEL_CRITICAL:
+            return log_level::k_critical;
+        case MONGOC_LOG_LEVEL_WARNING:
+            return log_level::k_warning;
+        case MONGOC_LOG_LEVEL_MESSAGE:
+            return log_level::k_message;
+        case MONGOC_LOG_LEVEL_INFO:
+            return log_level::k_info;
+        case MONGOC_LOG_LEVEL_DEBUG:
+            return log_level::k_debug;
+        case MONGOC_LOG_LEVEL_TRACE:
+            return log_level::k_trace;
+        default:
+            // TODO: MONGOCXX_UNREACHABLE (CXX-628)
+            std::abort();
+    }
+}
+
+void null_log_handler(::mongoc_log_level_t, const char *, const char *, void *) {
```

Comment:
This appears to be unused now. I think you should set it as the log handler in the instance::impl ctor for the case where the logger unique_ptr is null.